Change FilterCriteria to use Java8's ZonedDateTime

lpapazow opened the pull request:
As suggested in this [discussion](https://github.com/eclipse/smarthome/issues/594), changed FilterCriteria  class to use Java8's ZonedDateTime class, instead of java.util.Date. 
The methods that use java.util.Date are marked as deprecated. 
When converting java.util.Date to ZonedDateTime, the default timezone is used.

Signed-off-by: Lyubomir V. Papazov <lyubomir.papazov@MUSALA.COM>

maggu2810 commented:
Shouldn't we use the timezone provided by the `TimeZoneProvider` OSGi service instead?

lpapazow commented:
Sure, using this service seems way more appropriate.

However, I'm not sure what would be the best way to access the `TimeZoneProvider `service in the `FilterCriteria` class. I am currently using the default timezone in two of `FilterCriteria`'s methods:

- `setBeginDate(Date beginDate)`
- `setEndDate(Date beginDate)`
 which I've marked as deprecated in this PR.

Do you think it would be appropriate to use the default timezone in these two methods? Alternatively, I could get the `TimeZoneProvider` service from the BundleContext, use that service to get the timezone and keep the timezone into a static variable in the `FilterCriteria` class. Or maybe there is some better way to do this that I can't think of?

maggu2810 commented:
As you marked it deprecated and there is another function that uses a better function argument, it is okay for me to use the system default one at this place.

Doesn't the default timezone / zone id be used [here](https://github.com/eclipse/smarthome/pull/4738/files#diff-e3fb993b083190530555d6efd632e55cR246) too? As this resource is a component, you could depend on the resp. service.

triller-telekom commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/FilterCriteria.java:
```diff
@@ -1,155 +1,227 @@
-    }
-
-}
+/**
```
I just noticed that the diff of this file looks rather large and also lines that are not changed are marked as changed. I think this might be a an issue with line endings ore file encoding because the indentation with spaces looks alright.

Could you please run the Eclipse formater again on this file?

triller-telekom commented on bundles/model/org.eclipse.smarthome.model.persistence/OSGI-INF/persistenceextension.xml:
```diff
@@ -16,4 +16,5 @@
 <scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" immediate="true" name="org.eclipse.smarthome.model.persistence.extensions">
    <implementation class="org.eclipse.smarthome.model.persistence.extensions.PersistenceExtensions"/>
    <reference bind="setPersistenceServiceRegistry" cardinality="1..1" interface="org.eclipse.smarthome.core.persistence.PersistenceServiceRegistry" name="PersistenceServiceRegistry" policy="static" unbind="unsetPersistenceServiceRegistry"/>
+   <reference bind="setTimeZoneProvider" cardinality="1..1" interface="org.eclipse.smarthome.core.i18n.TimeZoneProvider" name="TimeZoneProvider" policy="static" unbind="unsetTimeZoneProvider"/>
```
Since we have switched to DS annotations since a while, could you please convert this xml file to annotations since you are touching it anyway. And while you are at it, would you please be so kind and also convert the `persistencemanager.xml` next to it too? Because then you can place a `.gitignore` in the `OSGI-INF` folder with `/*.xml`.

triller-telekom commented on bundles/model/org.eclipse.smarthome.model.persistence/src/org/eclipse/smarthome/model/persistence/extensions/PersistenceExtensions.java:
```diff
@@ -55,11 +59,19 @@ public PersistenceExtensions() {
         PersistenceExtensions.registry = registry;
     }
-
+    
```
There are some spaces in this line, please remove them.

triller-telekom commented on bundles/model/org.eclipse.smarthome.model.persistence/src/org/eclipse/smarthome/model/persistence/extensions/PersistenceExtensions.java:
```diff
@@ -144,7 +156,7 @@ public static HistoricItem historicState(Item item, AbstractInstant timestamp, S
             QueryablePersistenceService qService = (QueryablePersistenceService) service;
             FilterCriteria filter = new FilterCriteria();
-            filter.setEndDate(timestamp.toDate());
+            filter.setEndDate(ZonedDateTime.ofInstant(timestamp.toDate().toInstant(), TimeZone.getDefault().toZoneId()));
```
Should this not be `timeZoneProvider.getTimeZone()`?

triller-telekom commented on bundles/ui/org.eclipse.smarthome.ui/OSGI-INF/chartprovider.xml:
```diff
@@ -20,4 +20,5 @@
    <service>
       <provide interface="org.eclipse.smarthome.ui.chart.ChartProvider"/>
    </service>
+   <reference bind="setTimeZoneProvider" cardinality="1..1" interface="org.eclipse.smarthome.core.i18n.TimeZoneProvider" name="TimeZoneProvider" policy="static" unbind="unsetTimeZoneProvider"/>
```
I would like to kindly ask you if you can also convert this to DS-annotations but the `OSGI-INF` folder contains 4 files, so I leave it to you if you want to convert all of them or at least just this one and ignore this single created file from the Eclipse IDE in the `.gitignore` file.

lpapazow commented:
As requested, I converted all of the `.xml` files in `org.eclipse.smarthome.ui/OSGI-INF` and `org.eclipse.smarthome.model.persistence/OSGI-INF` (6 files totally) to annotations,  deleted all `.xml` files and added `.gitignore` files in these two folders.

Also, in the previous version of this PR the `PersistenceExtensionsTest` were breaking with NPE because they were dependent on a `timeZoneProvider` object which wasn't provided by the previous test configuration, so I added `org.eclipse.equinox.ds` to the configuration of the `tycho-surefire-plugin` in `org.eclipse.smarthome.model.persistence.tests/pom.xml`.

maggu2810 commented:
@triller-telekom Do you want to have a final look?

kaikreuzer commented on bundles/ui/org.eclipse.smarthome.ui/src/main/java/org/eclipse/smarthome/ui/internal/chart/defaultchartprovider/DefaultChartProvider.java:
```diff
@@ -106,6 +119,16 @@ protected void activate() {
         logger.debug("Available themes for default chart provider: {}", themeNames);
     }
 
+    @Reference(cardinality=ReferenceCardinality.MANDATORY, policy=ReferencePolicy.STATIC)
```
Mandatory-Static is the default and thus does not require to be explicitly stated here, right?

maggu2810 replied:
Yes, its already implicit set, so no need to do it explicit.

kaikreuzer commented:
Oops, that comment was too late... As this appears at multiple places in this PR, maybe you could create a new PR for the cleanup - thanks!
